lib/commons/color/get-stroke-colors-from-shadows.js

```diff
@@ -0,0 +1,90 @@
+import Color from './color';
+
+/** Magic numbers **/
+// Alpha value to use when text shadows are offset between .5px and 1.5px
+const SHADOW_STROKE_ALPHA = 0.54;
+// Shadows offset by less than this are not visible enough no matter how much you stack them
+const VISIBLE_SHADOW_MIN_PX = 0.5;
+// Shadows offset by more than this have full opacity
+const OPAQUE_STROKE_OFFSET_MIN_PX = 1.5;
+
+const edges = ['top', 'right', 'bottom', 'left'];
+
+/**
+ * Work out which color(s) of an array of text shadows form a stroke around the text.
+ * @param {Array[]} testShadows Parsed test shadows (see color.parseTestShadow())
+ * @returns {Array[]} Array of colors
+ */
+export default function getStrokeColorsFromShadows(parsedShadows) {
+  const shadowMap = getShadowColorsMap(parsedShadows);
+  const shadowsByColor = Object.entries(shadowMap).map(([colorStr, sides]) => {
+    const edgeCount = edges.filter(side => sides[side].length !== 0).length;
+    return { colorStr, sides, edgeCount };
+  });
+
+  // Skip shadows that cover too much of the text to be ignored, but not enough to be tested
+  if (shadowsByColor.some(({ edgeCount }) => edgeCount > 1 && edgeCount < 4)) {
+    return null;
+  }
+
+  return shadowsByColor
+    .map(shadowGroupToColor)
+    .filter(shadow => shadow !== null);
+}
+
+/**
+ * Create a map of colors to the sides they are on
+ */
+function getShadowColorsMap(parsedShadows) {
+  const colorMap = {};
+  for (const { colorStr, pixels } of parsedShadows) {
+    colorMap[colorStr] ??= { top: [], right: [], bottom: [], left: [] };
+    const borders = colorMap[colorStr];
+    const [offsetX, offsetY] = pixels;
+
+    if (offsetX > VISIBLE_SHADOW_MIN_PX) {
+      borders.right.push(pixels);
+    } else if (offsetX < -VISIBLE_SHADOW_MIN_PX) {
+      borders.left.push(pixels);
+    }
+    if (offsetY > VISIBLE_SHADOW_MIN_PX) {
+      borders.bottom.push(pixels);
+    } else if (offsetY < -VISIBLE_SHADOW_MIN_PX) {
+      borders.top.push(pixels);
+    }
+  }
+  return colorMap;
+}
+
+/**
+ * Using colorStr and thickness of sides, create a color object
+ */
+function shadowGroupToColor({ colorStr, sides, edgeCount }) {
+  if (edgeCount !== 4) {
```

Comment:
It feels off that this check is still here when the caller is now *also* doing an overlapping edgeCount check. Maybe more clear to move this check out of `shadowGroupToColor` and update the `.filter` on L32 to filter out based on edgeCount before mapping this function, instead of mapping this function and then filtering based on its return value.